### ISSUE TITLE
IoTHubCommunication Exception

### ISSUE THREAD

**Issue by kgalic** [Azure/azure-iot-sdk-csharp#1430]
I am using IoT Edge 1.0.9.3 running on the Debian distribution of Linux. I have several python modules and only one .NET based module. 
On some devices this module works perfectly, but on other devices I get this weird exception - it is weird because it only happens with .NET based module - python modules are working. 

```
---> Microsoft.Azure.Devices.Client.Exceptions.IotHubCommunicationException: Transient network error occurred, please retry.
---> System.Net.Internals.SocketExceptionFactory+ExtendedSocketException (00000005, 0xFFFDFFFF): Name or service not known
   at System.Net.Dns.InternalGetHostByName(String hostName)
   at System.Net.Dns.ResolveCallback(Object context)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw(Exception source)
   at System.Net.Dns.HostResolutionEndHelper(IAsyncResult asyncResult)
   at System.Net.Dns.EndGetHostAddresses(IAsyncResult asyncResult)
   at System.Net.Dns.<>c.<GetHostAddressesAsync>b__25_1(IAsyncResult asyncResult)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Azure.Devices.Client.Transport.Mqtt.MqttTransportHandler.OpenAsyncInternal(CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.Transport.Mqtt.MqttTransportHandler.OpenAsync(CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.Transport.ProtocolRoutingDelegatingHandler.OpenAsync(CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.Transport.ErrorDelegatingHandler.<>c__DisplayClass23_0.<<ExecuteWithErrorHandlingAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Azure.Devices.Client.Transport.ErrorDelegatingHandler.ExecuteWithErrorHandlingAsync[T](Func`1 asyncOperation)
```

I am using the default retry logic, and I used the VSCode template for .NET module without any modification when it comes to connecting via ModuleClient. 

As there is obviously different behavior for different devices with the same module, my feeling is that this is either bug, or something is not documented properly. 
Please let me know what is the problem here. 

UPTDATE:
I am using the latest version of DeviceClient SDK: 1.26.0

Furthermore, I noticed if I remove this from my create options that it works:
```
"NetworkingConfig": {
    "EndpointsConfig": {
      "host": {}
    }
  },
  "HostConfig": {
    "NetworkMode": "host"
  }
```
And also, if I add this to any python module, it also breaks. 
But I need this setting from other reasons. How to keep this, but also make the module client working and connecting properly.

**Comment by timtay-microsoft** [comment]
Do you see this exception consistently for your .NET modules? Or is it just occasionally?

**Comment by kgalic** [comment]
Today I have learned that it happens only with the custom OS image provided by third party. And it happens every time, with .NET and python modules if I include create options as described above. 

However, this does not happen with official Raspian image. 

I’ve spent some time already, trying various options, for example, having the same setting on EdgeHub create options, but it did not help.
Also, I found on other git thread that this exception happens if there is a proxy setting in place, that in that case it does the DNS lookup, but I could not find any proxy setting.

Furthermore, other modules without this setting, work perfectly..

Any ideas how to avoid this exception but to keep network settings in create options?

**Comment by kgalic** [comment]
Update:

The same exception I have observed as well in the Edge Hub - this time it is not the custom image:


```
System.TimeoutException: The operation timed out. ---> Microsoft.Azure.Devices.Client.Exceptions.IotHubCommunicationException: Transient network error occurred, please retry. ---> System.OperationCanceledException: The operation was canceled.
   at System.Threading.CancellationToken.ThrowOperationCanceledException()
   at System.Threading.SemaphoreSlim.WaitUntilCountOrTimeoutAsync(TaskNode asyncWaiter, Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.Transport.Mqtt.MqttTransportHandler.SendTwinRequestAsync(Message request, String rid, CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.Transport.Mqtt.MqttTransportHandler.SendTwinGetAsync(CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.Transport.ErrorDelegatingHandler.ExecuteWithErrorHandlingAsync[T](Func`1 asyncOperation)
   --- End of inner exception stack trace ---
   at Microsoft.Azure.Devices.Client.Transport.ErrorDelegatingHandler.ExecuteWithErrorHandlingAsync[T](Func`1 asyncOperation)
   at Microsoft.Azure.Devices.Client.Transport.RetryDelegatingHandler.<>c__DisplayClass24_0.<<SendTwinGetAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Azure.Devices.Client.Transport.RetryDelegatingHandler.SendTwinGetAsync(CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.InternalClient.GetTwinAsync()
   --- End of inner exception stack trace ---
   at Microsoft.Azure.Devices.Client.InternalClient.GetTwinAsync()
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient.InvokeFunc[T](Func`1 func, String operation, Boolean useForConnectivityCheck)
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient.InvokeFunc[T](Func`1 func, String operation, Boolean useForConnectivityCheck) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/ConnectivityAwareClient.cs:line 173
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.CloudProxy.GetTwinAsync() in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/CloudProxy.cs:line 115
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.CloudProxy.GetTwinAsync() in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/CloudProxy.cs:line 126
   at Microsoft.Azure.Devices.Edge.Hub.Core.Cloud.RetryingCloudProxy.ExecuteOperation[T](Func`2 func, String operation)
   at Microsoft.Azure.Devices.Edge.Hub.Core.Twin.CloudSync.<>c__DisplayClass4_0.<<GetTwin>b__0>d.MoveNext() in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/twin/CloudSync.cs:line 36
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Azure.Devices.Edge.Hub.Core.Twin.CloudSync.GetTwin(String id) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/twin/CloudSync.cs:line 33
<6> 2020-07-06 06:44:45.862 +00:00 [INF] - Entering connected state
<6> 2020-07-06 06:44:52.561 +00:00 [INF] - Updated reported properties for kgProdEdgeDevice3/AgentModule
<6> 2020-07-06 06:44:52.726 +00:00 [INF] - Started task to cleanup processed and stale messages for endpoint kgProdEdgeDevice3/RelayModule/input
<6> 2020-07-06 06:44:52.727 +00:00 [INF] - Cleaned up 0 messages from queue for endpoint kgProdEdgeDevice3/RelayModule/input and 0 messages from message store.
<6> 2020-07-06 06:45:03.099 +00:00 [INF] - Updated reported properties for kgProdEdgeDevice3/AgentModule
<6> 2020-07-06 06:45:10.493 +00:00 [INF] - Exiting connected state
<6> 2020-07-06 06:45:10.494 +00:00 [INF] - Entering unreachable state
<4> 2020-07-06 06:45:10.515 +00:00 [WRN] - Error getting twin for kgProdEdgeDevice3/CapacitiveMoistureModule
System.TimeoutException: The operation timed out. ---> Microsoft.Azure.Devices.Client.Exceptions.IotHubCommunicationException: Transient network error occurred, please retry. ---> System.OperationCanceledException: The operation was canceled.
   at System.Threading.CancellationToken.ThrowOperationCanceledException()
   at System.Threading.SemaphoreSlim.WaitUntilCountOrTimeoutAsync(TaskNode asyncWaiter, Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.Transport.Mqtt.MqttTransportHandler.SendTwinRequestAsync(Message request, String rid, CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.Transport.Mqtt.MqttTransportHandler.SendTwinGetAsync(CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.Transport.ErrorDelegatingHandler.ExecuteWithErrorHandlingAsync[T](Func`1 asyncOperation)
   --- End of inner exception stack trace ---
   at Microsoft.Azure.Devices.Client.Transport.ErrorDelegatingHandler.ExecuteWithErrorHandlingAsync[T](Func`1 asyncOperation)
   at Microsoft.Azure.Devices.Client.Transport.RetryDelegatingHandler.<>c__DisplayClass24_0.<<SendTwinGetAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Azure.Devices.Client.Transport.RetryDelegatingHandler.SendTwinGetAsync(CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.InternalClient.GetTwinAsync()
   --- End of inner exception stack trace ---
   at Microsoft.Azure.Devices.Client.InternalClient.GetTwinAsync()
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient.InvokeFunc[T](Func`1 func, String operation, Boolean useForConnectivityCheck)
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient.InvokeFunc[T](Func`1 func, String operation, Boolean useForConnectivityCheck) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/ConnectivityAwareClient.cs:line 173
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.CloudProxy.GetTwinAsync() in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/CloudProxy.cs:line 115
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.CloudProxy.GetTwinAsync() in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/CloudProxy.cs:line 126
   at Microsoft.Azure.Devices.Edge.Hub.Core.Cloud.RetryingCloudProxy.ExecuteOperation[T](Func`2 func, String operation)
   at Microsoft.Azure.Devices.Edge.Hub.Core.Twin.CloudSync.<>c__DisplayClass4_0.<<GetTwin>b__0>d.MoveNext() in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/twin/CloudSync.cs:line 36
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Azure.Devices.Edge.Hub.Core.Twin.CloudSync.GetTwin(String id) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/twin/CloudSync.cs:line 33
<6> 2020-07-06 06:45:10.675 +00:00 [INF] - Entering connected state
<6> 2020-07-06 06:45:13.786 +00:00 [INF] - Updated reported properties for kgProdEdgeDevice3/AgentModule
<6> 2020-07-06 06:45:21.152 +00:00 [INF] - Exiting connected state
<6> 2020-07-06 06:45:21.153 +00:00 [INF] - Entering unreachable state
<4> 2020-07-06 06:45:21.173 +00:00 [WRN] - Error getting twin for kgProdEdgeDevice3/RelayModule
System.TimeoutException: The operation timed out. ---> Microsoft.Azure.Devices.Client.Exceptions.IotHubCommunicationException: Transient network error occurred, please retry. ---> System.OperationCanceledException: The operation was canceled.
   at System.Threading.CancellationToken.ThrowOperationCanceledException()
   at System.Threading.SemaphoreSlim.WaitUntilCountOrTimeoutAsync(TaskNode asyncWaiter, Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.Transport.Mqtt.MqttTransportHandler.SendTwinRequestAsync(Message request, String rid, CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.Transport.Mqtt.MqttTransportHandler.SendTwinGetAsync(CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.Transport.ErrorDelegatingHandler.ExecuteWithErrorHandlingAsync[T](Func`1 asyncOperation)
   --- End of inner exception stack trace ---
   at Microsoft.Azure.Devices.Client.Transport.ErrorDelegatingHandler.ExecuteWithErrorHandlingAsync[T](Func`1 asyncOperation)
   at Microsoft.Azure.Devices.Client.Transport.RetryDelegatingHandler.<>c__DisplayClass24_0.<<SendTwinGetAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Azure.Devices.Client.Transport.RetryDelegatingHandler.SendTwinGetAsync(CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.InternalClient.GetTwinAsync()
   --- End of inner exception stack trace ---
   at Microsoft.Azure.Devices.Client.InternalClient.GetTwinAsync()
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient.InvokeFunc[T](Func`1 func, String operation, Boolean useForConnectivityCheck)
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient.InvokeFunc[T](Func`1 func, String operation, Boolean useForConnectivityCheck) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/ConnectivityAwareClient.cs:line 173
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.CloudProxy.GetTwinAsync() in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/CloudProxy.cs:line 115
   at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.CloudProxy.GetTwinAsync() in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/CloudProxy.cs:line 126
   at Microsoft.Azure.Devices.Edge.Hub.Core.Cloud.RetryingCloudProxy.ExecuteOperation[T](Func`2 func, String operation)
   at Microsoft.Azure.Devices.Edge.Hub.Core.Twin.CloudSync.<>c__DisplayClass4_0.<<GetTwin>b__0>d.MoveNext() in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/twin/CloudSync.cs:line 36
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Azure.Devices.Edge.Hub.Core.Twin.CloudSync.GetTwin(String id) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/twin/CloudSync.cs:line 33
<6> 2020-07-06 06:45:21.472 +00:00 [INF] - Entering connected state
<6> 2020-07-06 06:45:24.466 +00:00 [INF] - Updated reported properties for kgProdEdgeDevice3/AgentModule
<6> 2020-07-06 06:45:35.245 +00:00 [INF] - Updated reported properties for kgProdEdgeDevice3/AgentModule
<6> 2020-07-06 06:45:45.744 +00:00 [INF] - Updated reported properties for kgProdEdgeDevice3/AgentModule
<6> 2020-07-06 06:45:56.273 +00:00 [INF] - Updated reported properties for kgProdEdgeDevice3/AgentModule
```

Also, I noticed that this happens if I turn off the device, and turn on after a certain period of time. After the initial deployment, it works. 
Rebooting the device did not help.

**Comment by ancaantochi** [comment]
I noticed a similar issue from edgeAgent. It looks like the network is disconnected for about 2 minutes then it reconnects. edgeHub can reconnect to IotHub after those 2 minutes, but edgeAgent connection doesn't recover even after 24 hours:


System.TimeoutException: The operation timed out. ---> Microsoft.Azure.Devices.Client.Exceptions.IotHubCommunicationException: Transient network error occurred, please retry. ---> System.TimeoutException: The operation has timed out.
   at Microsoft.Azure.Devices.Client.Transport.Amqp.AmqpTransportHandler.RoundTripTwinMessage(AmqpTwinMessageType amqpTwinMessageType, TwinCollection reportedProperties, CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.Transport.Amqp.AmqpTransportHandler.SendTwinGetAsync(CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.Transport.ErrorDelegatingHandler.ExecuteWithErrorHandlingAsync[T](Func`1 asyncOperation)
   --- End of inner exception stack trace ---
   at Microsoft.Azure.Devices.Client.Transport.ErrorDelegatingHandler.ExecuteWithErrorHandlingAsync[T](Func`1 asyncOperation)
   at Microsoft.Azure.Devices.Client.Transport.RetryDelegatingHandler.<>c__DisplayClass24_0.<<SendTwinGetAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Azure.Devices.Client.Transport.RetryDelegatingHandler.SendTwinGetAsync(CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.InternalClient.GetTwinAsync()
   --- End of inner exception stack trace ---
   at Microsoft.Azure.Devices.Client.InternalClient.GetTwinAsync()
   at Microsoft.Azure.Devices.Edge.Agent.IoTHub.ModuleClient.GetTwinAsync() in /home/vsts/work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/ModuleClient.cs:line 86
   at Microsoft.Azure.Devices.Edge.Agent.IoTHub.ModuleClient.GetTwinAsync() in /home/vsts/work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/ModuleClient.cs:line 91
   at Microsoft.Azure.Devices.Edge.Agent.IoTHub.EdgeAgentConnection.<>c__DisplayClass30_0.<<GetTwinFromIoTHub>g__GetTwinFunc|0>d.MoveNext() in /home/vsts/work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/EdgeAgentConnection.cs:line 198
--- End of stack trace from previous location where exception was thrown ---

**Comment by kgalic** [comment]
any news on this one?

**Comment by abhipsaMisra** [comment]
@kgalic  From your experiments it looks like default options work as expected, but adding the following to create options causes it to break:
```json
"NetworkingConfig": {
    "EndpointsConfig": {
      "host": {}
    }
  },
  "HostConfig": {
    "NetworkMode": "host"
  }
```

Could you point to the initializers that you are using to create this edge module?
@ancaantochi Do you all see this when specifying these "create" options, or does it happen for some other scenario?

**Comment by kgalic** [comment]
I used VSCode template for the edge module            
```
            MqttTransportSettings mqtt = new MqttTransportSettings(TransportType.Mqtt_Tcp_Only);
            ITransportSettings[] settings = { mqtt };

            // Open a connection to the Edge runtime
            ModuleClient ioTHubModuleClient = await ModuleClient.CreateFromEnvironmentAsync(settings);
            await ioTHubModuleClient.OpenAsync();
```

**Comment by vinagesh** [comment]
@kgalic - This issue has been fixed in edge agent 1.0.9.4. The latest is 1.0.10 but if you do not wish to upgrade to this version, we highly recommend version 1.0.9.5. Closing this issue. Please reopen if the issue is not resolved.

**Comment by az-iot-builder-01** [comment]
@timtay-microsoft, @kgalic, @ancaantochi, @abhipsaMisra, @vinagesh, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)
